### PR TITLE
Next generation BunnyBus

### PR DESCRIPTION
**Pull request by cesarhq**
## Description

- node support for `8.x || 10.x || 12.x`
- removing `callbacks` in favor of `async/await` and native `promises`
- improved connection retries upon server disconnects
- removed dependencies of 3rd party flow control libraries
- updated dependencies 

## Types of changes

- [ ] Documentation only (no changes to either `lib/` or `test/` files)
- [ ] Bug fix (non-breaking change which fixes an issue. you didn't modify existing tests)
- [ ] New feature (non-breaking change which adds functionality. you added at least one new test)
- [x] Breaking change (fix or feature that would cause existing functionality to change. you had to modify existing tests.)

## Checklist:

- [ ] I have read the [**CONTRIBUTING**](https://github.com/xogroup/bunnybus/blob/master/.github/CONTRIBUTING.md) document.
- [x] My code follows the [Hapi.js style guide](https://github.com/hapijs/contrib/blob/master/Style.md).
- [x] I have updated the documentation as needed.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by cesarhq** [comment]
see also release draft

https://github.com/xogroup/bunnybus/releases

**Comment by cesarhq** [comment]
look what I did ma! @lamchakchan

File: lib/helpers/createTransactionId.js [hunk 1]
```diff
@@ -2,14 +2,9 @@
...
-const createTransactionId = (callback) => {
-
-    setImmediate(() => {
-
-        const hash = Crypto.randomBytes(20).toString('hex');
-
-        callback(null, hash);
-    });
-};
+const createTransactionId = () =>
```
Comment (dhinklexo) [inline lib/helpers/createTransactionId.js]: Why force this to be async?

File: lib/helpers/index.js [hunk 2]
```diff
@@ -1,20 +1,35 @@
 'use strict';
 
+const delay = (timeout) => {
+    return new Promise((resolve) => setTimeout(resolve, timeout));
+};
+
+const loop = async ({ condition, exec }) => {
```
Comment (dhinklexo) [inline lib/helpers/index.js]: This looks an awful lot like metaprogramming...

**Comment by cesarhq** [comment]
@dhinklexo @WesTyler did some clean up as PR comments and also was able to finally test reconnection attempts

**Comment by dhinklexo** [comment]
Fixed merge conflicts, going to merge this into a release branch so some other fixes can be added to it before we do the v4 release this represents

**Comment by ten-lac** [comment]
@dhinklexo , FYI, I rewrote this entire library recently.  It includes functionality around multiple connection/channel support.  Also supports different message delegation patterns.

https://github.com/tenna-llc/bunnybus

**Comment by dhinklexo** [comment]
@ten-lac Well you saved me a bunch of hours since I just sat down to make the same batch of fixes. Will check it out to see if we can just merge your fork back in as a v4.

**Comment by ten-lac** [comment]
> @ten-lac Well you saved me a bunch of hours since I just sat down to make the same batch of fixes. Will check it out to see if we can just merge your fork back in as a v4.

You should be able to as long as you roll back to the commit prior to the merge of this PR.  You'll have a lot of merge conflicts to deal with if you merge the latest from `tenna-llc/bunnybus` against this PR since it doesn't have the same lineage

**Pull request merged**